subsys/net/lib/async_socket/async_socket.c

```diff
@@ -0,0 +1,380 @@
+		if (entry->sock == sock) {
+			return entry;
+		}
+	}
+	return NULL;
+}
+
+static void rcv_callbacks_init(void)
+{
+	struct rcv_callbacks *entry, *end;
+
+	end = rcv_callback + ARRAY_SIZE(rcv_callback);
+	for (entry = &rcv_callback[0]; entry < end; entry++) {
+		entry->sock = INVALID_SOCK;
+	}
+}
+
+static int rcv_callback_register(int sock,
+				 void *buf, size_t max_len,
+				 async_recv_cb_t cb, void *cb_data)
+{
+	int status = 0;
+	struct rcv_callbacks *rcv_cb = NULL;
+
+	/* See if this sock already has a registered callback: */
+	rcv_cb = get_rcv_callback_slot(sock);
+	if (!rcv_cb) {
+		/* If not, see if we have room: */
+		if (num_rcv_cbs == MAX_RCV_CALLBACKS) {
+			/* No more room: */
+			NET_ERR("Increase CONFIG_NET_SOCKETS_POLL_MAX");
+		} else {
+			/* Find new slot for this socket: */
+			rcv_cb = get_rcv_callback_slot(INVALID_SOCK);
+		}
+	}
+
+	if (rcv_cb) {
+		rcv_cb->sock = sock;
+		rcv_cb->buf = buf;
+		rcv_cb->max_len = max_len;
+		rcv_cb->cb = cb;
+		rcv_cb->cb_data = cb_data;
+
+		num_rcv_cbs++;
+
+		/* Now, signal the signal_sock to restart the poll server,
+		 * so it can recreate the ufds[] list and start poll()
+		 * with the newly registered socket.
+		 */
+		status = async_server_restart();
+	} else {
+		status = -ENOMEM;
+	}
+
+	return status;
+}
+
+static void rcv_callback_deregister(short int sock)
+{
+	int status = 0;
+	struct rcv_callbacks *rcv_cb = NULL;
+
+	/* Find slot of this socket: */
+	rcv_cb = get_rcv_callback_slot(sock);
+	if (rcv_cb) {
+		rcv_cb->sock = INVALID_SOCK;
+		rcv_cb->buf = NULL;
+		rcv_cb->max_len = 0;
+		rcv_cb->cb = NULL;
+		rcv_cb->cb_data = NULL;
+
+		num_rcv_cbs--;
+
+		/* Now, signal the signal_sock to restart the poll server,
+		 * so it can recreate the ufds[] list and start poll()
+		 * *without* the newly un-registered socket.
+		 */
+		status = async_server_restart();
+	}
+}
+
+/* Fill the poll() ufds argument from the rcv_callback list: */
+static void rcv_callbacks_to_ufds(struct pollfd *fds)
+{
+	struct rcv_callbacks *entry, *end;
+	int fds_index = 0;
+
+	end = rcv_callback + ARRAY_SIZE(rcv_callback);
+	for (entry = &rcv_callback[0], fds_index = 0; entry < end; entry++) {
+		if (entry->sock != INVALID_SOCK) {
+			fds[fds_index].fd = entry->sock;
+			fds[fds_index].events = POLLIN;
+			fds_index++;
+		}
+	}
+}
+
+/*
+ * Call poll() in a loop, waiting for registered async sockets
+ * to get signalled; if so, call their registered receive callbacks.
+ *
+ * Currently, this only handles receive.  Send, accept and other
+ * events are not handled.
+ *
+ * Problem: how to break out of a blocking poll() to reset
+ * the list of events?	(Without using a finite timeout -
+ * which is not good from a power perspective).
+ * In Linux, we could use eventfd() to add an fd to the poll()
+ * list of fds, to enable breaking out of the poll() when
+ * a new fd is added/removed.
+ * But, Zephyr's poll() API only supports sockets for now, and Zephyr
```

Comment:
Right, but Zephyr has k_poll() which can do that. Yep, it would be Zephyr-specific implementation, but that's similar to how eventfd() is Linux-specific implementation (it's not POSIX).

But per above, I don't think that we should even go there, we just shouldn't incur extra thread and related synchronization overheads in the first place.